src/inmanta/ast/statements/call.py

```diff
@@ -261,13 +262,17 @@ def __init__(self, queue_scheduler, resolver, result: ResultVariable, requires,
         self.ready(self)
 
     def execute(self):
-        requires = {k: v.get_value() for (k, v) in self.requires.items()}
+        requires = {k: v.get_value() for (k, v) in self.base_requires.items()}
         try:
             self.function.resume(requires, self.resolver, self.queue_scheduler, self.result)
             self.done = True
         except UnsetException as e:
             LOGGER.debug("Unset value in python code in plugin %s %s.%s.", self.function.function, e.instance, e.attribute)
-            self.waitfor(e.get_result_variable())
+            # Don't handle it here!
+            # This exception is used by the scheduler to re-queue the unit
+            # If it is handled here, the re-queueing can not be done,
+            # leading to very subtle errors such as #2787
+            raise
```

Comment:
It's still not entirely clear to me why it can't be handled here. This could just call `requeue_with_additional_requires` with the same effect, couldn't it?
(Not saying that it should, just trying to understand the subtleties of the issue.)

Reply:
No, in the `try: except` block in scheduler is another remove statement after this one. 

So catching it here allows that discard to be execute, which should not happen. 

```python
              try:
                    next.execute()
                    all_statements.discard(next)
                    count = count + 1
               except UnsetException as e:
                    # some statements don't know all their dependencies up front,...
                    next._requeue_with_additional_requires(e, e.get_result_variable())
```

Reply:
I see, thanks.